src/util/treadmill.rs

```diff
@@ -6,111 +6,192 @@ use crate::util::ObjectReference;
 
 use super::object_enum::ObjectEnumerator;
 
+/// A data structure for recording objects in the LOS.
+///
+/// It is divided into the nursery and the mature space, and each of them is further divided into
+/// the from-space and the to-space.
+///
+/// All operations are protected by a single mutex [`TreadMill::sync`].
 pub struct TreadMill {
-    from_space: Mutex<HashSet<ObjectReference>>,
-    to_space: Mutex<HashSet<ObjectReference>>,
-    collect_nursery: Mutex<HashSet<ObjectReference>>,
-    alloc_nursery: Mutex<HashSet<ObjectReference>>,
+    sync: Mutex<TreadMillSync>,
+}
+
+/// The synchronized part of [`TreadMill`]
+#[derive(Default)]
+struct TreadMillSync {
+    nursery: SpacePair,
+    mature: SpacePair,
+}
+
+/// A pair of from and two spaces.
+#[derive(Default)]
+struct SpacePair {
+    from_space: HashSet<ObjectReference>,
+    to_space: HashSet<ObjectReference>,
+}
+
+/// Used by [`TreadMill::enumerate_objects`] to determine what to do to objects in the from-spaces.
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+pub(crate) enum FromSpacePolicy {
```

Comment:
I don't think we need a type so the caller can describe what they need to `TreadMill`. They can just do
```rust
self.treadmill.enumerate_objects(...);
self.treadmill.enumerate_from_space_objects(...);
assert!(self.treadmill.is_from_space_empty());
```